### PR TITLE
Add Dod/Gcch url for trouter/registrar

### PR DESCRIPTION
**Pull request by CHILIU-MSFT**
Add Dod/Gcch url for trouter/registrar
Add url auto-detect logic

### CODE REVIEW

File: sdk/communication/azure-communication-chat/src/main/java/com/azure/android/communication/chat/implementation/notifications/NotificationUtils.java [hunk 1]
```diff
@@ -218,12 +241,15 @@ public static OffsetDateTime parseEpochTime(Long epochMilli) {
...
+        String skypeId = decodeSkypeIdFromJwtToken(skypeToken);
+        String prefix = skypeId.substring(0, skypeId.indexOf(":"));
 
-    public static JSONArray getJSONArray(JSONObject payload, String property) throws JSONException {
-        return new JSONArray(payload.getString(property));
+        if (SKYPE_ID_CLOUD_TYPES_MAPPING.containsKey(prefix)) {
+            return SKYPE_ID_CLOUD_TYPES_MAPPING.get(prefix);
+        }
+
+        return  CloudType.Public;
```
Comment (vcolin7) [inline sdk/communication/azure-communication-chat/src/main/java/com/azure/android/communication/chat/implementation/notifications/NotificationUtils.java:252]: nit;
```suggestion
        return CloudType.Public;
```

File: sdk/communication/azure-communication-chat/src/main/java/com/azure/android/communication/chat/implementation/notifications/NotificationUtils.java [hunk 2]
```diff
@@ -218,12 +241,15 @@ public static OffsetDateTime parseEpochTime(Long epochMilli) {
...
 
-    public static JSONObject getJSONObject(JSONObject payload, String property) throws JSONException {
-        return new JSONObject(payload.getString(property));
-    }
+    public static CloudType getUserCloudTypeFromSkypeToken(String skypeToken) {
+        String skypeId = decodeSkypeIdFromJwtToken(skypeToken);
+        String prefix = skypeId.substring(0, skypeId.indexOf(":"));
 
-    public static JSONArray getJSONArray(JSONObject payload, String property) throws JSONException {
-        return new JSONArray(payload.getString(property));
```
Comment (vcolin7) [inline sdk/communication/azure-communication-chat/src/main/java/com/azure/android/communication/chat/implementation/notifications/NotificationUtils.java:226]: Is this no longer required?

File: sdk/communication/azure-communication-chat/src/main/java/com/azure/android/communication/chat/implementation/notifications/fcm/RegistrarClient.java [hunk 3]
```diff
@@ -232,4 +238,25 @@ private boolean awaitOnLatch(CountDownLatch latch) {
             throw logger.logExceptionAsError(new RuntimeException("Operation didn't complete within " + timeoutInMin + " minutes"));
         }
     }
+
+    private String getRegistrarServiceUrl(String skypeUserToken) {
+        CloudType cloudType = NotificationUtils.getUserCloudTypeFromSkypeToken(skypeUserToken);
+        String registrarUrl = PUSHNOTIFICATION_REGISTRAR_SERVICE_URL;
```
Comment (vcolin7) [inline sdk/communication/azure-communication-chat/src/main/java/com/azure/android/communication/chat/implementation/notifications/fcm/RegistrarClient.java:244]: We can move this to the default case below.

File: sdk/communication/azure-communication-chat/src/main/java/com/azure/android/communication/chat/implementation/notifications/signaling/CommunicationSignalingClient.java [hunk 4]
```diff
@@ -201,11 +215,31 @@ public String load() {
...
 
-    private void start(Context context, ISkypetokenProvider skypetokenProvider) {
+    private void start(Context context, ISkypetokenProvider skypetokenProvider, String skypeUserToken) {
         if (this.isRealtimeNotificationsStarted) {
             return;
         }
 
+        CloudType cloudType = NotificationUtils.getUserCloudTypeFromSkypeToken(skypeUserToken);
+        String trouterUrl = TROUTER_HOSTNAME;
+        String registrarUrl = TROUTER_REGISTRATION_HOSTNAME;
```
Comment (vcolin7) [inline sdk/communication/azure-communication-chat/src/main/java/com/azure/android/communication/chat/implementation/notifications/signaling/CommunicationSignalingClient.java:225]: We can move these two lines to the default case below.

File: sdk/communication/azure-communication-chat/src/main/java/com/azure/android/communication/chat/implementation/notifications/NotificationUtils.java (hunk 2 above)
Comment (CHILIU-MSFT) [reply sdk/communication/azure-communication-chat/src/main/java/com/azure/android/communication/chat/implementation/notifications/NotificationUtils.java:226]: No one calls these 2 functions. This was used to parse incoming notification payload. This functionality is replaced by Jackson & annotations.

**Review by vcolin7 — APPROVED** [review]
LGTM :)

**Pull request merged**